sympy/solvers/solvers.py

```diff
@@ -2804,26 +2804,42 @@ def nsolve(*args, **kwargs):
         # e.g., issue 11768
 
         f = lambdify(fargs, f, modules)
-        return sympify(findroot(f, x0, **kwargs))
-
-    if len(fargs) > f.cols:
-        raise NotImplementedError(filldedent('''
-            need at least as many equations as variables'''))
-    verbose = kwargs.get('verbose', False)
-    if verbose:
-        print('f(x):')
-        print(f)
-    # derive Jacobian
-    J = f.jacobian(fargs)
-    if verbose:
-        print('J(x):')
-        print(J)
-    # create functions
-    f = lambdify(fargs, f.T, modules)
-    J = lambdify(fargs, J, modules)
-    # solve the system numerically
-    x = findroot(f, x0, J=J, **kwargs)
-    return Matrix(x)
+        x = sympify(findroot(f, x0, **kwargs))
+
+    else:
+        if len(fargs) > f.cols:
+            raise NotImplementedError(filldedent('''
+                need at least as many equations as variables'''))
+        verbose = kwargs.get('verbose', False)
+        if verbose:
+            print('f(x):')
+            print(f)
+        # derive Jacobian
+        J = f.jacobian(fargs)
+        if verbose:
+            print('J(x):')
+            print(J)
+        # create functions
+        f = lambdify(fargs, f.T, modules)
+        J = lambdify(fargs, J, modules)
+        # solve the system numerically
+        x = findroot(f, x0, J=J, **kwargs)
+        x = Matrix(x)
+
+    #
+    # done, result is ready to be returned and is stored in x
+    ###########################################################################
+
+    # check if user wants result to be returned as a dictionary
+    as_dict = kwargs.get('dict', False)
+    if as_dict:
+        # x won't be iterable (neither list, tuple nor Matrix) if there is only one variable in fargs, therefore exception TypeError will be raised
```

Comment:
> Can we check this condition directly instead of try/except?

I got the impression, that try/except would be the best solution as it doesn't really matter if the return value is a Matrix or tuple or list or something else - it just depends on whether or not the result is iterable. This should be the only attribute which decides if the function needs to create the dictionary with `dict(zip(fargs, [sympify(xi) for xi in x]))` or `dict([(fargs, x)])`. Is that bad style?

> In fact, why not split this across the above if/else block

Sorry, I thought it would be cleaner (?) to complete the computation of the result first, and then to format the return value. I tried to make it similar to how the `solve` function handles it. Should I rewrite this with two `if as_dict:` statements at the end of each if/else block?